File: backend/app/schemas/plot_planning.py
from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, field_validator, model_validator

PlotUnitStatus = Literal["draft", "active", "completed", "archived"]
PlotPlanRevisionStatus = Literal["draft", "active", "stale", "blocked", "superseded", "completed"]
PlanningDecisionStatus = Literal["pending", "resolved", "superseded"]
PlanningDecisionSource = Literal["during_generation", "during_review"]
DraftRevisionStatus = Literal["candidate", "applied", "superseded"]


class AuthorFoundationOut(BaseModel):
    id: int
    novel_id: int
    outline: str
    current_intent: str
    stage_goal: str
    constraints_json: dict[str, Any]
    version: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class AuthorFoundationUpdateRequest(BaseModel):
    outline: Optional[str] = None
    current_intent: Optional[str] = None
    stage_goal: Optional[str] = None
    constraints_json: Optional[dict[str, Any]] = None


class AuthorFoundationRevisionOut(BaseModel):
    id: int
    novel_id: int
    foundation_id: int
    version: int
    snapshot_json: dict[str, Any]
    change_reason: str
    created_at: datetime

    model_config = {"from_attributes": True}


class PlotUnitCreateRequest(BaseModel):
    title: str
    scope_type: str
    start_position: int
    end_position: int
    author_goal: str = ""
    start_state: str = ""
    end_state: str = ""


class PlotUnitOut(BaseModel):
    id: int
    novel_id: int
    title: str
    scope_type: str
    start_position: int
    end_position: int
    author_goal: str
    start_state: str
    end_state: str
    foundation_revision_id: int
    status: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PlotPlanRevisionOut(BaseModel):
    id: int
    novel_id: int
    plot_unit_id: int
    version: int
    foundation_revision_id: int
    based_on_published_chapter_id: Optional[int] = None
    plan_json: dict[str, Any]
    change_reason: str
    status: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PlanningDecisionOut(BaseModel):
    id: int
    novel_id: int
    plot_unit_id: Optional[int] = None
    plot_plan_revision_id: Optional[int] = None
    chapter_brief_id: Optional[int] = None
    writing_run_id: Optional[int] = None
    source: str
    status: str
    conflict_summary: str
    evidence_json: dict[str, Any]
    options_json: list
    recommended_index: int
    recommendation_reason: str
    impact_scope_json: dict[str, Any]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class DecisionOptionOut(BaseModel):
    label: str
    consequence: str


class ChooseDecisionRequest(BaseModel):
    option_index: Optional[int] = None
    custom_intent: Optional[str] = None

    @model_validator(mode="after")
    def require_option_or_custom_intent(self):
        if self.option_index is None and not self.custom_intent:
            raise ValueError("必须提供 option_index 或 custom_intent")
        if self.option_index is not None and self.custom_intent:
            raise ValueError("不能同时提供 option_index 和 custom_intent")
        return self


class DraftRevisionOut(BaseModel):
    id: int
    novel_id: int
    writing_run_id: Optional[int] = None
    parent_revision_id: Optional[int] = None
    decision_id: Optional[int] = None
    base_content: str
    candidate_content: str
    scope_json: dict[str, Any]
    diff_json: dict[str, Any]
    reason: str
    status: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/plotPlanning.ts
export type PlotPlanStatus = 'draft' | 'active' | 'stale' | 'blocked' | 'superseded' | 'completed'
export type DecisionSource = 'during_generation' | 'during_review'
export type DecisionStatus = 'pending' | 'resolved' | 'superseded'
export type DraftRevisionStatus = 'candidate' | 'applied' | 'superseded'

export interface AuthorFoundation {
  id: number
  novel_id: number
  outline: string
  current_intent: string
  stage_goal: string
  constraints_json: Record<string, unknown>
  version: number
  created_at: string
  updated_at: string
}

export interface AuthorFoundationUpdate {
  outline?: string
  current_intent?: string
  stage_goal?: string
  constraints_json?: Record<string, unknown>
}

export interface AuthorFoundationRevision {
  id: number
  novel_id: number
  foundation_id: number
  version: number
  snapshot_json: Record<string, unknown>
  change_reason: string
  created_at: string
}

export interface PlotUnit {
  id: number
  novel_id: number
  title: string
  scope_type: string
  start_position: number
  end_position: number
  author_goal: string
  start_state: string
  end_state: string
  foundation_revision_id: number
  status: string
  created_at: string
  updated_at: string
}

export interface PlotUnitCreate {
  title: string
  scope_type: string
  start_position: number
  end_position: number
  author_goal: string
  start_state: string
  end_state: string
}

export interface PlotPlanRevision {
  id: number
  novel_id: number
  plot_unit_id: number
  foundation_revision_id: number
  based_on_published_chapter_id: number | null
  version: number
  plan_json: Record<string, unknown>
  change_reason: string
  status: PlotPlanStatus
  created_at: string
  updated_at: string
}

export interface PlanningDecision {
  id: number
  novel_id: number
  plot_unit_id: number | null
  plot_plan_revision_id: number | null
  chapter_brief_id: number | null
  writing_run_id: number | null
  source: DecisionSource
  status: DecisionStatus
  conflict_summary: string
  evidence_json: Record<string, unknown>
  options_json: DecisionOption[]
  recommended_index: number
  recommendation_reason: string
  impact_scope_json: Record<string, unknown>
  selected_option_index?: number
  custom_intent?: string
  created_at: string
  updated_at: string
}

export interface DecisionOption {
  label: string
  action: string
  consequence: string
}

export interface DraftRevision {
  id: number
  novel_id: number
  writing_run_id: number | null
  parent_revision_id: number | null
  decision_id: number | null
  base_content: string
  candidate_content: string
  scope_json: Record<string, unknown>
  diff_json: Record<string, unknown>
  reason: string
  status: DraftRevisionStatus
  created_at: string
  updated_at: string
}

export interface ChooseDecisionRequest {
  option_index?: number
  custom_intent?: string
}

export interface DecisionResolution {
  decision: PlanningDecision
  new_plan: PlotPlanRevision | null
  draft_revision: DraftRevision
}

export interface ReviewWritingRunResponse {
  decision: PlanningDecision | null
  review_issues: unknown[]
}
